proofs/hol_light/arm/proofs/mlkem_poly_basemul_acc_montgomery_cached_k2.ml

```diff
@@ -0,0 +1,369 @@
+      0x4e6081cb;       (* arm_SMLAL2_VEC Q11 Q14 Q0 16 *)
+      0xd10005ad;       (* arm_SUB X13 X13 (rvalue (word 1)) *)
+      0xb5fff9ed;       (* arm_CBNZ X13 (word 2096956) *)
+      0x4e71c065;       (* arm_SMULL2_VEC Q5 Q3 Q17 16 *)
+      0x0e6081cc;       (* arm_SMLAL_VEC Q12 Q14 Q0 16 *)
+      0x0e648157;       (* arm_SMLAL_VEC Q23 Q10 Q4 16 *)
+      0x3d800409;       (* arm_STR Q9 X0 (Immediate_Offset (word 16)) *)
+      0x4e648145;       (* arm_SMLAL2_VEC Q5 Q10 Q4 16 *)
+      0x4e4b598b;       (* arm_UZP2 Q11 Q12 Q11 16 *)
+      0x4e5b3a69;       (* arm_ZIP1 Q9 Q19 Q27 16 128 *)
+      0x0e6f81b7;       (* arm_SMLAL_VEC Q23 Q13 Q15 16 *)
+      0x4e6f81a5;       (* arm_SMLAL2_VEC Q5 Q13 Q15 16 *)
+      0x3c820409;       (* arm_STR Q9 X0 (Postimmediate_Offset (word 32)) *)
+      0x0e768397;       (* arm_SMLAL_VEC Q23 Q28 Q22 16 *)
+      0x4e768385;       (* arm_SMLAL2_VEC Q5 Q28 Q22 16 *)
+      0x4e451ae9;       (* arm_UZP1 Q9 Q23 Q5 16 *)
+      0x4e629d29;       (* arm_MUL_VEC Q9 Q9 Q2 16 128 *)
+      0x4e608125;       (* arm_SMLAL2_VEC Q5 Q9 Q0 16 *)
+      0x0e608137;       (* arm_SMLAL_VEC Q23 Q9 Q0 16 *)
+      0x4e455ae9;       (* arm_UZP2 Q9 Q23 Q5 16 *)
+      0x4e4b7925;       (* arm_ZIP2 Q5 Q9 Q11 16 128 *)
+      0x4e4b3929;       (* arm_ZIP1 Q9 Q9 Q11 16 128 *)
+      0x3d800405;       (* arm_STR Q5 X0 (Immediate_Offset (word 16)) *)
+      0x3c820409;       (* arm_STR Q9 X0 (Postimmediate_Offset (word 32)) *)
+      0x6d4027e8;       (* arm_LDP D8 D9 SP (Immediate_Offset (iword (&0))) *)
+      0x6d412fea;       (* arm_LDP D10 D11 SP (Immediate_Offset (iword (&16))) *)
+      0x6d4237ec;       (* arm_LDP D12 D13 SP (Immediate_Offset (iword (&32))) *)
+      0x6d433fee;       (* arm_LDP D14 D15 SP (Immediate_Offset (iword (&48))) *)
+      0x910103ff;       (* arm_ADD SP SP (rvalue (word 64)) *)
+      0xd65f03c0        (* arm_RET X30 *)
+    ];;
+
+let pmull = define
+  `pmull (x0: 16 word) (x1 : 16 word) (y0 : 16 word) (y1 : 16 word) =
+      word_add (word_mul ((word_sx x1) : 32 word) (word_sx y1))
+               (word_mul ((word_sx x0) : 32 word) (word_sx y0))`;;
+
+let pmull_acc2 = define
+    `pmull_acc2 (x00: 16 word) (x01 : 16 word) (y00 : 16 word) (y01 : 16 word)
+          (x10: 16 word) (x11 : 16 word) (y10 : 16 word) (y11 : 16 word) =
+      word_add (pmull x10 x11 y10 y11) (pmull x00 x01 y00 y01)`;;
+
+let montred = define
+   `montred (x : 32 word) =
+      word_subword (
+         word_add (
+           word_mul (
+             (word_sx : 16 word -> 32 word) (
+               word_mul (
+                 word_subword x (0,16)
+               ) (word 3327)
+             )
+           )
+           (word 3329)
+         ) x
+      ) (16, 16)`;;
+
+let pmul_acc2 = define
+    `pmul_acc2 (x00: 16 word) (x01 : 16 word) (y00 : 16 word) (y01 : 16 word)
+          (x10: 16 word) (x11 : 16 word) (y10 : 16 word) (y11 : 16 word) =
+      montred (pmull_acc2 x00 x01 y00 y01 x10 x11 y10 y11)`;;
```

Comment:
It is somewhat confusing that here y00 and y01 are actually b0, and b1*twiddle in the even case, and b1 and b0 in the even case. 
In case you have to touch this again, maybe it makes sense to rename and reorder the arguments so that two coeffients next to each other are multiplied?

Reply:
Ok!